continued work on `get_labeled_inventory`

BenPortner opened the pull request:
continued work on `get_labeled_inventory`
- [x] implemented helper functions to bulk load metadata from databases (addressing #15)
- [x] implemented switch to return labeled inventory data in wide or long format (addressing #13) 
- [x] implemented tests for long format
- [x] implement helper function to label characterized inventories (addressing #14 )

open issues:
- it might be worthwhile to store metadata in a cache to avoid loading it from the hard drive each time `get_labeled_inventory` is called. What is your opinion on this @cmutel ?
- should these helpers be accessible via `lca.inventory.to_dataframe()` and `lci.characterized_inventory.to_dataframe()`?

Commit:
continued work on `get_labeled_inventory`:
- implemented helper to bulk load metadata from databases
- implemented switch to return labeled inventory data in wide or long format
- implemented tests for long format

cmutel commented on bw2analyzer/lci.py:
```diff
@@ -1,14 +1,60 @@
 
 
-def get_labeled_inventory(lca: LCA) -> pd.DataFrame:
+def _get_dependent_database_names(lca: LCA) -> List[str]:
```
Duplicates existing functionality, and assumes presence of `remapping_dicts`. Please use [Database.find_dependents](https://github.com/brightway-lca/brightway2-data/blob/8627519fd9a01cc6504b10278f7948a627e58b63/bw2data/backends/base.py#L135) or [Data.find_graph_dependents](https://github.com/brightway-lca/brightway2-data/blob/8627519fd9a01cc6504b10278f7948a627e58b63/bw2data/backends/base.py#L161) with the functional unit. I know this is a bit more work, but it is worth it.

cmutel replied:
You can merge if you just want to get this done with, but at some point we should eliminate duplicate functionality, it will cause pain in the future when things change.

BenPortner replied:
No worries, I will eliminate it before merging. What is your opinion on the metadata caching (see #15)?

cmutel commented on bw2analyzer/lci.py:
```diff
@@ -1,14 +1,60 @@
+
+
+def _load_database_metadata(
+    lca: LCA, cols: Union[None, List[str], str] = None
```
Best practice to use `Optional` instead of `Union[None` (I just learned this myself).

cmutel commented on bw2analyzer/lci.py:
```diff
@@ -1,14 +1,60 @@
+    # load metadata and store in one dataframe
+    df = pd.DataFrame()
+    for db in databases:
+        df = df.append(pd.DataFrame(Database(db)), ignore_index=True)
```
Kind of amazing that this works.

BenPortner replied:
Agreed :)

cmutel commented on bw2analyzer/lci.py:
```diff
@@ -1,14 +1,60 @@
+        cols = df.columns.tolist()
+    existing_cols = [c for c in cols if c in df]
+    df = df[existing_cols]
+    # set index
```
Maybe unnecessary :)

BenPortner replied:
Only unnecessary if `usecols = 'all'`. In all other cases, it's necessary.

cmutel replied:
Sorry, I meant the comment :)

BenPortner replied:
Oh! Hahaha, true.

cmutel commented on bw2analyzer/lci.py:
```diff
@@ -1,14 +1,60 @@
     Args:
         * *lca* (bw2calc.LCA): LCA object whose life cycle inventory has been calculated previously.
+        * *wide_format* (bool): Whether to return a pd.DataFrame in wide format (biosphere x technosphere)
+            or a pd.Series in long format ((biosphere x technosphere) x 1)
```
Long is also called tidy in the R world, at least. Documentation here should be clearer, I can imagine what this means but am not sure.

BenPortner replied:
Okay. Do you think it would help to refer to this wikipedia article? https://en.wikipedia.org/wiki/Wide_and_narrow_data

cmutel replied:
I think this is fine for the code, but the docs should have a nice presentation of the dataframe with the indices in pretty html.

cmutel commented on bw2analyzer/lci.py:
```diff
@@ -18,17 +64,63 @@ def get_labeled_inventory(lca: LCA) -> pd.DataFrame:
+            .reset_index(drop=True)
+        )
+
+        # rename columns to avoid duplicate names for row and col
```
Can you briefly describe why you think this is preferred versus adding another level to the multiindex?

BenPortner replied:
Not sure, what exactly you are referring to. The `id` is already a separate column. Here, I am dropping the index to be able to concatenate row and column metadata.

cmutel replied:
Sorry, maybe I am just off in my own world, but can't multiindices be hierarchical? I would find it much cleaner to have ('foo', 'name') than 'foo_name'. But maybe pandas doesn't actually work like this. I have to admit I am having flashbacks to when others have done these types of string manipulations and I had to reverse engineer them, it wasn't fun...

BenPortner replied:
Ah, you are referring to the MultiIndex names. Well, no the names can't be hierarchical. It might be possible to assign tuples as names but then standard methods like `query` won't work. I would strongly advice against it.

cmutel reviewed (commented):
A few comments. I am still not sure why the same code couldn't be applied to the characterized inventory with hardly any modification, it has the same dimensions.

cmutel commented on bw2analyzer/lci.py:
```diff
@@ -1,14 +1,60 @@
+    # load metadata and store in one dataframe
+    df = pd.DataFrame()
+    for db in databases:
+        df = df.append(pd.DataFrame(Database(db)), ignore_index=True)
```
It's faster to do this:

```python
def get_db(db):
       keys = ['activity', 'activity type', 'authors', 'classifications', 'code',
               'comment', 'database', 'filename', 'flow', 'id', 'location', 'name',
               'parameters', 'production amount', 'reference product', 'synonyms',
               'type', 'unit']
       return pd.DataFrame([{key: obj.get(key) for key in keys} for obj in Database(db)])
```

2.7 versus 3.8 seconds. However, your approach is easier and more flexible, so fine with me.

BenPortner replied:
That's interesting. I thought iterating over the database was slow (see #15) but it seems another part was. Perhaps the `get_activity`?

I am trying to avoid pre-determining the keys because users might have custom keys in their databases (e.g. Exiobase uses "compartments" where Ecoinvent uses "categories"). Is it still faster if we ask for `obj.keys()` instead of the fixed list?

cmutel replied:
It's fine, your question is the exact reason we shouldn't try to hard code things, you never know what users are up to...

Commit:
lci tagger overhaul:
- use `Database.find_graph_dependents` instead of custom method
- use typing.Optional instead of Union[None...]
- remove unnecessar comment #set index

Commit:
Merge remote-tracking branch 'origin/master' into lci-label-contd

Commit:
attempt better description of `lci.get_labeled_inventory`

Commit:
added helper to label characterized inventories

BenPortner commented:
Almost ready to merge. Requesting your opinion on the last two open issues in the description above @cmutel (caching, accessibility via bw2calc).
